doc/case-studies/xycloans/xycloans_monitor.tla

```diff
@@ -0,0 +1,138 @@
+\* integer division with rounding up
+div_ceil(a, b) == a + (b - 1) \div b
+\* integer division with rounding down
+div_floor(a, b) == a \div b
+\* @type: ($env, Str) => Int;
+token_balance(env, a) == get_or_else(env.storage.token_persistent.Balance, a, 0)
+\* @type: ($env, Str) => Int;
+old_token_balance(env, a) == get_or_else(env.old_storage.token_persistent.Balance, a, 0)
+
+\* @type: $tx => Bool;
+initialize(tx) ==
+    LET call == AsInitialize(tx.call) IN
+    /\ IsInitialize(tx.call)
+    /\ reverts_with(tx, tx.env.old_storage.self_instance.TokenId /= "")
+    /\ succeeds_with(tx, call.token = XLM_TOKEN_SAC_TESTNET)
+    /\ succeeds_with(tx, tx.env.storage.self_instance.TokenId = call.token)
+    \* these conditions are not required by a monitor, but needed to avoid spurious generated values
+    /\ succeeds_with(tx,
+        tx.env.storage.self_instance.FeePerShareUniversal = tx.env.old_storage.self_instance.FeePerShareUniversal)
+    /\ last_tx' = tx
+    /\ shares' = [ addr \in {} |-> 0 ]
+    /\ total_shares' = 0
+    /\ fee_per_share_universal' = 0    
+
+\* @type: $tx => Bool;
+deposit(tx) ==
+    LET call == AsDeposit(tx.call)
+        token == tx.env.storage.self_instance.TokenId
+        new_shares == [ shares EXCEPT ![call.from] = @ + call.amount ]
+    IN
+    /\ IsDeposit(tx.call)
+    \* the pool has received `amount` tokens
+    /\  LET a == tx.env.current_contract_address IN
+        succeeds_with(tx, token_balance(tx.env, a) = old_token_balance(tx.env, a) + call.amount)
+    \* `from` received `amount` shares
+    /\  LET b == get_or_else(tx.env.storage.self_persistent.Balance, call.from, 0)
+            old_b == get_or_else(tx.env.old_storage.self_persistent.Balance, call.from, 0)
+        IN
+        /\ succeeds_with(tx, new_shares[call.from] = b)
+        /\ succeeds_with(tx, b = old_b + call.amount)
+    \* update the monitor state
+    /\ last_tx' = tx
+    /\ shares' = new_shares
+    /\ total_shares' = total_shares + call.amount
+    /\ UNCHANGED fee_per_share_universal
+
+\* @type: $tx => Bool;
+borrow(tx) ==
+    LET call == AsBorrow(tx.call)
+        expected_fee == div_ceil(call.amount * STROOP, 12500000000)
+        expected_fee_per_share_universal ==
+            tx.env.old_storage.self_instance.FeePerShareUniversal
+                + div_floor(expected_fee * STROOP, total_shares)
+    IN
+    /\ IsBorrow(tx.call)
+    \* `FeePerShareUniversal` has been updated correctly
+    /\ succeeds_with(tx,
+           expected_fee_per_share_universal = tx.env.storage.self_instance.FeePerShareUniversal)
+    \* the receiver paid the expected fee to the pool
+    /\ succeeds_with(tx, old_token_balance(tx.env, call.receiver_id) >= call.amount)
+    /\ LET a == call.receiver_id IN
+       succeeds_with(tx, token_balance(tx.env, a) = old_token_balance(tx.env, a) - call.amount)
+    /\ LET a == tx.env.current_contract_address IN
+       succeeds_with(tx, token_balance(tx.env, a) = old_token_balance(tx.env, a) + call.amount)
```

Comment:
I don't like that I have to substitute `a` in my brain. We should find a better name.

Also, repeating it is confusing, as it refers to different addresses.